### PR TITLE
Wait for sent keys to be picked up by the keyboard hook.

### LINKED ISSUE #15822: No feedback when using Backspace while typing in Kannada language using Nudi 6.1 software
### Steps to reproduce:
Received via email

Using Nudi6.1 software. (you can download from below link.)
https://kagapa.in/kannada/sites/default/files/downloads/Nudi-6.1/Nudi_6.1_setup.exe

Type any text and then use backspace to remove any characters.

### Actual behavior:

With NVDA's Keyboard setting "Handle keys from other applications" checked, user receives the following error:
"71 hotkeys have been received in the last 1812ms.
Do you want to continue?
yes  no
(see #MaxHotkeysPerInterval in the help file)"

With this option disabled, the action works, however with no feedback from NVDA.

(Regarding this setting, the user guide does note that "Certain users may wish to turn this off, such as those typing Vietnamese with the UniKey typing software as it will cause incorrect character input")

The user advised that when trying the same thing with NVDA 2022.4, they do not receive the error.  But when using NVDA 2023.1 or later, they do.

### Expected behavior:

Ideally the user should get feedback about deleted characters, and no error.

#### NVDA version:

NVDA 2023.3 installed

#### Windows version:

Windows 10.0.19045 Build 19045

#### Name and version of other software in use when reproducing the issue:

for typing Kannada, we use Nudi6.1 software. (you can download from below link.)
https://kagapa.in/kannada/sites/default/files/downloads/Nudi-6.1/Nudi_6.1_setup.exe

#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.

Also tested with NVDA 2022.4 and NVDA 2023.1

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

### PR DESCRIPTION
**Pull request by jcsteh**
### Link to issue number:
Fixes #15822.

### Summary of the issue:
With "Handle keys from other applications" enabled, NVDA interferes with certain software which intercepts and re-transmits keyboard commands such as Nudi6.1. For example, this makes it impossible to use backspace while this software is running.

### Description of user facing changes
Backspace now works correctly when using Nudi6.1 with NVDA's "Handle keys from other applications" setting enabled.

### Description of development approach
Before #14708, we delayed slightly when sending keys, which would mean that any keys sent by an app which captured a key sent by NVDA would be ignored during that period. After #14708, we only ignore any keys received while NVDA is sending keys, but that doesn't include any keys captured and sent afterward by, for example, Nudi6.1. This could result in a loop where NVDA kept receiving the key it sent (re-transmitted by the other app) and sending it again.

To fix this, we now explicitly wait for the key sent by NVDA (e.g. backspace) to be received by the keyboard hook before we stop ignoring keys and thus return from KeyboardInputGesture.send(). This means that if another application intercepts this key and re-transmits it, we will wait for that re-transmission and ignore it. We use a kernel event so that the notification from the keyboard hook can be handled as quickly as possible without being dependent on the system timer resolution.

There is a chance that a key will be intercepted and never re-sent. To deal with this, we wait for a maximum timeout of 10 ms. This means that in the best case scenario, nothing intercepts a key sent by NVDA and we return almost immediately. In the worst case scenario, we wait between 10 and 30 ms (depending on the system timer resolution), which is no worse than the situation before #14708.

### Testing strategy:
I don't have Nudi6.1, but the reporter confirmed this fix works in https://github.com/nvaccess/nvda/issues/15822#issuecomment-1845190338.

### Known issues with pull request:
None.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: include/liblouis
Comment (LeonarddeR) [inline include/liblouis:1]: COuld you undo this?

File: source/keyboardHandler.py [hunk 1]
```diff
@@ -35,13 +35,16 @@
...
 import threading
+import winKernel
 
 if typing.TYPE_CHECKING:
 	from NVDAObjects import NVDAObject  # noqa: F401
 	from watchdog import WatchdogObserver
 
 _watchdogObserver: typing.Optional["WatchdogObserver"] = None
 ignoreInjected=False
+_lastInjectedKeyUp: typing.Tuple[int, int] = None
```
Comment (seanbudd) [inline source/keyboardHandler.py:46]: I believe this is the standard syntax for python 3.11 now

```suggestion
_lastInjectedKeyUp: Optional[tuple[int, int]] = None
```

File: source/keyboardHandler.py [hunk 2]
```diff
@@ -35,13 +35,16 @@
...
+_injectionDoneEvent: int = None
```
Comment (seanbudd) [inline source/keyboardHandler.py:47]:
```suggestion
_injectionDoneEvent: Optional[int] = None
```

File: source/keyboardHandler.py [hunk 3]
```diff
@@ -567,7 +574,11 @@ def executeScript(self, script):
 		# Now actually execute the script.
 		super().executeScript(script)
 
+	#: The maximum amount of time (in ms) to wait for keys injected by NVDA to be
+	#: received by NVDA.
+	_INJECTION_WAIT_TIMEOUT: int = 10
```
Comment (seanbudd) [inline source/keyboardHandler.py:579]: Added new line for lint

```suggestion
	_INJECTION_WAIT_TIMEOUT: int = 10

```

File: user_docs/en/changes.t2t [hunk 4]
```diff
@@ -13,6 +13,8 @@ What's New in NVDA
 
 
 == Bug Fixes ==
+- Backspace now works correctly when using Nudi6.1 with NVDA's "Handle keys from other applications" setting enabled. (#15822, @jcsteh)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:16]:
```suggestion
- Backspace now works correctly when using Nudi 6.1 with NVDA's "Handle keys from other applications" setting enabled. (#15822, @jcsteh)
```

**Review by seanbudd — COMMENTED** [review]
Thanks @jcsteh, changes generally look good to me

File: source/keyboardHandler.py (hunk 1 above)
Comment (seanbudd) [reply source/keyboardHandler.py:46]: alternatively

```suggestion
_lastInjectedKeyUp: tuple[int, int] | None = None
```

**Review by seanbudd — APPROVED** [review]

**Pull request merged**